cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/AppMetadataStore.java

```diff
@@ -573,41 +553,20 @@ private ProgramRunStatus recordProgramStop(ProgramId programId, String pid, long
     return runStatus;
   }
 
-  private List<RunRecordMeta> getRuns(ProgramId programId, String pid) {
-    ImmutableSet.Builder<MDSKey> keySet = ImmutableSet.<MDSKey>builder().add(
-      getProgramKeyBuilder(TYPE_RUN_RECORD_STARTING, programId).add(pid).build(),
-      getProgramKeyBuilder(TYPE_RUN_RECORD_STARTED, programId).add(pid).build(),
-      getProgramKeyBuilder(TYPE_RUN_RECORD_SUSPENDED, programId).add(pid).build(),
-      getProgramKeyBuilder(TYPE_RUN_RECORD_COMPLETED, programId).add(pid).build());
-    // Get version-less record of type TYPE_RUN_RECORD_COMPLETED only if upgrade is not complete and
-    // programId has default version. Because upgrade is only done for record type TYPE_RUN_RECORD_COMPLETED in
-    // one transaction, there won't be duplicated records for the same program run
-    if (!upgradeComplete.get() && ApplicationId.DEFAULT_VERSION.equals(programId.getVersion())) {
-      keySet.add(getVersionLessProgramKeyBuilder(TYPE_RUN_RECORD_COMPLETED, programId).add(pid).build());
-    }
-    return get(keySet.build(), RunRecordMeta.class);
-  }
-
   /**
    * Checks whether the existing run record metas of a given program run are in a state for
    * the program run to transition into the given run status.
    *
-   * @param existingRecords the existing run record metas of the given program run
+   * @param existing the existing run record meta of the given program run
    * @param programId id of the program
    * @param pid run id
    * @param sourceId the source id of the current program status
    * @param recordType the type of record corresponding to the current status
    * @param status the status that the program run is transitioning into
    * @return {@code true} if the program run is allowed to persist the given status, {@code false} otherwise
    */
-  private boolean validateExistingRecords(List<RunRecordMeta> existingRecords, ProgramId programId, String pid,
+  private boolean validateExistingRecords(RunRecordMeta existing, ProgramId programId, String pid,
```

Comment:
annotate existing as nullable and change the method name to `validateExistingRecord` since there's only one record now